src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+
+
+class DalMethodRequest:
+    """
+    Mutable request state. Middleware can modify this.
+    """
+
+    def __init__(
+            self, ctx: DataAccessContext, path: Tuple[PathSegment, ...], args: Tuple[Any, ...], kwargs: Dict
+    ):
+        self.ctx = ctx
+        self.path = path
+        self.args = args
+        self.kwargs = kwargs
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, request: DalMethodRequest, path: Optional[Tuple[PathSegment]] = None):
+        self._request = request
+        self._path = path
+
+
+@dataclass
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name: str
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta: dict
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+
+class DalMethodRequester:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+                TODO: Consider using a custom type for the handler to represent
+                      or document the interface to that callable.
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> DalMethodRequester:
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+
+class MethodMiddleware:
+    """
+    Base class used primarily for type checking and to provide
+    an interface for a method middleware.
```

Comment:
We were trying to introduce a pattern here that method middleware must be this type. We don't have to though if you don't like it. I'd like to avoid cryptic runtime/testtime errors because somehow a callable is a dict.

Reply:
I'm surprised [typing](https://docs.python.org/3/library/typing.html#functions-and-decorators) doesn't have some way to verify something matches a protocol. I know Python doesn't do run time checking, but it seems like `typing` should at least off a built in way to say "does this thing match this protocol I defined with `typing`?". At least I can't find a way. Looking for a standard library way so we don't have to add anything. Maybe we have to wait until anticipate supports typing.

All we need to check for now then is that it's callable. A class and a function/method is a callable. The advantage of allow that is you can have either a class, a single function, or a function that is a factory that eventually returns a middleware.

The base class isn't really needed. Classes are only used to make it easier to extend them, but to the middleware handler, they're just callables.